File: sdk-python/obsvr/presidio.py
"""Presidio integration (parity with sdk-typescript/src/policy/presidio.ts).

NLP-level PII detection and redaction via the Presidio analyzer
and anonymizer services. Every function is failure-safe: network errors and
timeouts return empty results / None so the caller falls back to the
built-in regex scanner. Entity mappings and typed placeholders mirror the
TypeScript SDK exactly, so both SDKs produce identical detected_types and
redacted output for the same Presidio deployment.
"""

import json
import re
from typing import Any, Dict, List, Optional
from urllib.request import Request, urlopen

# Presidio entity type -> internal PII label (parity with presidio.ts)
PRESIDIO_TO_LABEL: Dict[str, str] = {
    "PERSON": "name",
    "EMAIL_ADDRESS": "email",
    "US_SSN": "ssn",
    "PHONE_NUMBER": "phone",
    "IP_ADDRESS": "ip_address",
    "CREDIT_CARD": "credit_card",
    "LOCATION": "location",
    "US_BANK_NUMBER": "bank_account",
    "IBAN_CODE": "iban",
    "MEDICAL_LICENSE": "medical",
    "NRP": "national_id",
    "DATE_TIME": "date",
}

# Typed placeholders sent to the anonymizer per entity type
ENTITY_PLACEHOLDERS: Dict[str, str] = {
    "PERSON": "[REDACTED_PERSON]",
    "EMAIL_ADDRESS": "[REDACTED_EMAIL]",
    "US_SSN": "[REDACTED_SSN]",
    "PHONE_NUMBER": "[REDACTED_PHONE]",
    "IP_ADDRESS": "[REDACTED_IP]",
    "CREDIT_CARD": "[REDACTED_CC]",
    "LOCATION": "[REDACTED_LOCATION]",
}

_WORD_START = re.compile(r"\b[a-z]")


def _normalize_for_ner(text: str) -> str:
    """Capitalize word starts so spaCy NER catches lowercase proper nouns.

    Case-only change: character positions are identical to the original, so
    analyzer spans apply to the original text as-is.
    """
    return _WORD_START.sub(lambda m: m.group(0).upper(), text)


def _post_json(url: str, payload: Dict[str, Any], timeout_s: float) -> Optional[Any]:
    try:
        req = Request(
            url,
            data=json.dumps(payload).encode("utf-8"),
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        resp = urlopen(req, timeout=timeout_s)
        return json.loads(resp.read().decode("utf-8"))
    except Exception:
        return None


def _analyze(
    text: str, analyzer_url: str, timeout_s: float
) -> Optional[List[Dict[str, Any]]]:
    """Call /analyze. ``None`` when the analyzer did not answer, a list when it
    did — including the empty list, which means it looked and found nothing.

    Those two used to be the same value, and the sameness was load-bearing in
    the wrong direction: a timed-out analyzer was indistinguishable from a
    clean scan, so a caller credited a detector that never ran and a redactor
    returned the text unchanged as though there had been nothing to remove.
    The 500ms default makes that failure routine rather than exotic.
    """
    data = _post_json(
        f"{analyzer_url}/analyze",
        {"text": _normalize_for_ner(text), "language": "en"},
        timeout_s,
    )
    return data if isinstance(data, list) else None


def presidio_scan(
    text: str, analyzer_url: str, timeout_s: float = 0.5
) -> Dict[str, Any]:
    """Scan text with the Presidio analyzer.

    Returns ``{"detected_types": [...], "answered": bool}`` using internal
    labels. ``answered`` is False when the analyzer could not be reached, so a
    caller can tell "this detector found nothing" from "this detector did not
    run" — the difference between attributing a verdict to Presidio and
    claiming it.
    """
    results = _analyze(text, analyzer_url, timeout_s)
    if results is None:
        return {"detected_types": [], "answered": False}
    seen: List[str] = []
    for r in results:
        label = PRESIDIO_TO_LABEL.get(r.get("entity_type", ""))
        if label and label not in seen:
            seen.append(label)
    return {"detected_types": seen, "answered": True}


def presidio_redact_text(
    text: str,
    analyzer_url: str,
    anonymizer_url: str,
    timeout_s: float = 0.5,
) -> Optional[str]:
    """Redact one string via analyze + anonymize.

    Returns the anonymized string, the original when nothing was detected,
    or None on failure (caller should fall back to regex redaction).
    """
    results = _analyze(text, analyzer_url, timeout_s)
    if results is None:
        # The analyzer did not answer. Returning ``text`` here read as "there
        # was nothing to redact" and handed the caller its own input back as a
        # redacted copy.
        return None
    if not results:
        return text
    # Per-entity replace anonymizers with typed placeholders. Presidio
    # anonymizer expects the key "anonymizers", NOT "operators".
    anonymizers = {
        r["entity_type"]: {"type": "replace", "new_value": ENTITY_PLACEHOLDERS[r["entity_type"]]}
        for r in results
        if r.get("entity_type") in ENTITY_PLACEHOLDERS
    }
    data = _post_json(
        f"{anonymizer_url}/anonymize",
        {"text": text, "analyzer_results": results, "anonymizers": anonymizers},
        timeout_s,
    )
    if isinstance(data, dict) and isinstance(data.get("text"), str):
        return data["text"]
    return None


def presidio_redact_kwargs(
    kwargs: Dict[str, Any],
    analyzer_url: str,
    anonymizer_url: str,
    timeout_s: float = 0.5,
) -> None:
    """Walk structured LLM request kwargs and redact each text node in place.

    Falls back to the built-in regex redactor per node on Presidio failure.
    Handles system (Anthropic), messages[].content (OpenAI/Anthropic, string
    or parts[]), and contents[].parts[].text (Gemini).
    """
    from .policy import redact_builtin_pii

    def _redact(text: str) -> str:
        out = presidio_redact_text(text, analyzer_url, anonymizer_url, timeout_s)
        return out if out is not None else redact_builtin_pii(text)

    if isinstance(kwargs.get("system"), str):
        kwargs["system"] = _redact(kwargs["system"])

    if isinstance(kwargs.get("messages"), list):
        for msg in kwargs["messages"]:
            if not isinstance(msg, dict):
                continue
            if isinstance(msg.get("content"), str):
                msg["content"] = _redact(msg["content"])
            elif isinstance(msg.get("content"), list):
                for part in msg["content"]:
                    if isinstance(part, dict) and isinstance(part.get("text"), str):
                        part["text"] = _redact(part["text"])

    if isinstance(kwargs.get("contents"), list):
        for content in kwargs["contents"]:
            if isinstance(content, dict) and isinstance(content.get("parts"), list):
                for part in content["parts"]:
                    if isinstance(part, dict) and isinstance(part.get("text"), str):
                        part["text"] = _redact(part["text"])

File: sdk-typescript/src/policy/presidio.ts
/**
 * Presidio Integration for the SDK
 *
 * SDK-local helpers for NLP-level PII detection and redaction via the
 * Presidio analyzer and anonymizer services.
 *
 * Detection returns an explicit `answered` bit so callers can distinguish an
 * empty result from an unavailable analyzer. Redaction returns null on any
 * unavailable stage; a caller applying an NLP-only redaction must fail closed
 * rather than fall back to a regex tier that cannot locate that type.
 *
 * @packageDocumentation
 */

import { redactBuiltinPii } from './hook.js';

// ── Entity mappings ───────────────────────────────────────────────────────────

/** Map Presidio entity type → our internal PII label */
const PRESIDIO_TO_LABEL: Record<string, string> = {
  PERSON:           'name',
  EMAIL_ADDRESS:    'email',
  US_SSN:           'ssn',
  PHONE_NUMBER:     'phone',
  IP_ADDRESS:       'ip_address',
  CREDIT_CARD:      'credit_card',
  LOCATION:         'location',
  US_BANK_NUMBER:   'bank_account',
  IBAN_CODE:        'iban',
  MEDICAL_LICENSE:  'medical',
  NRP:              'national_id',
  DATE_TIME:        'date',
};

/** PII labels the built-in regex tier cannot locate. */
export const NLP_ONLY_PII_TYPES: ReadonlySet<string> = new Set([
  'name',
  'address',
  'person',
  'location',
  'medical',
  'national_id',
]);

/** Typed placeholders sent to the Presidio anonymizer per entity type */
const ENTITY_PLACEHOLDERS: Record<string, string> = {
  PERSON:        '[REDACTED_PERSON]',
  EMAIL_ADDRESS: '[REDACTED_EMAIL]',
  US_SSN:        '[REDACTED_SSN]',
  PHONE_NUMBER:  '[REDACTED_PHONE]',
  IP_ADDRESS:    '[REDACTED_IP]',
  CREDIT_CARD:   '[REDACTED_CC]',
  LOCATION:      '[REDACTED_LOCATION]',
};

// ── Internal helpers ──────────────────────────────────────────────────────────

function buildAbortSignal(timeoutMs: number): AbortSignal {
  return AbortSignal.timeout(timeoutMs);
}

/**
 * Capitalize the first letter of every lowercase-starting word so spaCy NER
 * recognizes names regardless of input casing (e.g. "bob" → "Bob").
 * Only the first character of each word is changed, so character positions
 * remain identical to the original - analyzer spans can be applied as-is.
 */
function normalizeForNer(text: string): string {
  return text.replace(/\b[a-z]/g, c => c.toUpperCase());
}

/** Call /analyze on Presidio; reports an unanswered result on any error. */
async function analyzeText(
  text: string,
  analyzerUrl: string,
  timeoutMs: number,
): Promise<{
  answered: boolean;
  results: Array<{ entity_type: string; start: number; end: number; score: number }>;
}> {
  try {
    const res = await fetch(`${analyzerUrl}/analyze`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      // Send case-normalized text so spaCy NER catches lowercase proper nouns.
      // Positions in the response still map to the original text (case-only change).
      body:    JSON.stringify({ text: normalizeForNer(text), language: 'en' }),
      signal:  buildAbortSignal(timeoutMs),
    });
    if (!res.ok) return { answered: false, results: [] };
    const data = await res.json();
    if (!Array.isArray(data)) return { answered: false, results: [] };
    return {
      answered: true,
      results: data as Array<{ entity_type: string; start: number; end: number; score: number }>,
    };
  } catch {
    return { answered: false, results: [] };
  }
}

/** Call /anonymize on Presidio; returns null on any error. */
async function anonymizeText(
  text: string,
  analyzerResults: Array<{ entity_type: string; start: number; end: number; score: number }>,
  anonymizerUrl: string,
  timeoutMs: number,
): Promise<string | null> {
  try {
    // Build per-entity replace anonymizers using our typed placeholders.
    // Presidio anonymizer expects the key "anonymizers", NOT "operators".
    const anonymizers: Record<string, { type: string; new_value: string }> = {};
    for (const r of analyzerResults) {
      if (ENTITY_PLACEHOLDERS[r.entity_type]) {
        anonymizers[r.entity_type] = {
          type:      'replace',
          new_value: ENTITY_PLACEHOLDERS[r.entity_type],
        };
      }
    }

    const res = await fetch(`${anonymizerUrl}/anonymize`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({ text, analyzer_results: analyzerResults, anonymizers }),
      signal:  buildAbortSignal(timeoutMs),
    });
    if (!res.ok) return null;
    const data = (await res.json()) as { text?: string };
    return data.text ?? null;
  } catch {
    return null;
  }
}

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Scan a text string with the Presidio analyzer.
 * `answered` distinguishes a healthy empty result from timeout/error.
 */
export async function presidioScan(
  text: string,
  analyzerUrl: string,
  timeoutMs = 500,
): Promise<{ detected_types: string[]; answered: boolean }> {
  const analyzed = await analyzeText(text, analyzerUrl, timeoutMs);
  const types = [
    ...new Set(
      analyzed.results
        .map(r => PRESIDIO_TO_LABEL[r.entity_type])
        .filter((t): t is string => t !== undefined),
    ),
  ];
  return { detected_types: types, answered: analyzed.answered };
}

/**
 * Redact a single text string via Presidio analyze + anonymize.
 * Returns the anonymized string, or null on any failure. Callers may use the
 * built-in fallback only when no NLP-only redaction is required.
 */
export async function presidioRedactText(
  text: string,
  analyzerUrl: string,
  anonymizerUrl: string,
  timeoutMs = 500,
): Promise<string | null> {
  const analyzed = await analyzeText(text, analyzerUrl, timeoutMs);
  if (!analyzed.answered) return null;
  if (analyzed.results.length === 0) return text;
  return anonymizeText(text, analyzed.results, anonymizerUrl, timeoutMs);
}

/**
 * Walk structured LLM request args and redact each text node with Presidio.
 * Falls back to redactBuiltinPii per node on Presidio failure unless an
 * NLP-only redaction is required, in which case it throws so the caller can
 * refuse the provider call.
 *
 * Handles:
 * - req.system          (string) - Anthropic system prompt
 * - req.messages[].content (string | parts[]) - OpenAI / Anthropic
 * - req.contents[].parts[].text - Gemini structured
 */
export async function presidioRedactArgs(
  args: unknown,
  analyzerUrl: string,
  anonymizerUrl: string,
  timeoutMs = 500,
  requireNlpRedaction = false,
): Promise<void> {
  if (!args || typeof args !== 'object') return;
  const req = args as Record<string, unknown>;
  let changedByPresidio = false;

  const redactText = async (text: string): Promise<string> => {
    const redacted = await presidioRedactText(
      text,
      analyzerUrl,
      anonymizerUrl,
      timeoutMs,
    );
    if (redacted === null) {
      if (requireNlpRedaction) {
        throw new Error('Presidio did not answer while applying an NLP-only redaction');
      }
      return redactBuiltinPii(text);
    }
    if (redacted !== text) changedByPresidio = true;
    return redactBuiltinPii(redacted);
  };

  // Anthropic system prompt
  if (typeof req.system === 'string') {
    req.system = await redactText(req.system);
  }

  // OpenAI / Anthropic messages[]
  if (Array.isArray(req.messages)) {
    for (const msg of req.messages as Array<Record<string, unknown>>) {
      if (typeof msg.content === 'string') {
        msg.content = await redactText(msg.content);
      } else if (Array.isArray(msg.content)) {
        for (const part of msg.content as Array<Record<string, unknown>>) {
          if (typeof part.text === 'string') {
            part.text = await redactText(part.text);
          }
        }
      }
    }
  }

  // Gemini contents[].parts[].text
  if (Array.isArray(req.contents)) {
    for (const content of req.contents as Array<Record<string, unknown>>) {
      if (Array.isArray(content.parts)) {
        for (const part of content.parts as Array<Record<string, unknown>>) {
          if (typeof part.text === 'string') {
            part.text = await redactText(part.text);
          }
        }
      }
    }
  }

  if (requireNlpRedaction && !changedByPresidio) {
    throw new Error('Presidio did not remove the detected NLP-only PII type');
  }
}
